sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -39,137 +64,178 @@ public sealed class ResourceIdentifier : IEquatable<ResourceIdentifier>, ICompar
 #pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
         public ResourceIdentifier(string resourceId)
         {
-            ResourceIdentifier id = Create(resourceId);
-            Init(id.Parent, id.ResourceType, id.Name, id.IsProviderResource);
-            _stringValue = resourceId;
-        }
+            Argument.AssertNotNullOrEmpty(resourceId, nameof(resourceId));
 
-        private ResourceIdentifier(ResourceIdentifier? parent, ResourceType resourceType, string resourceName, bool idProviderResource)
-        {
-            Init(parent ?? Root, resourceType, resourceName, idProviderResource);
-        }
-#pragma warning restore CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+            if (resourceId.Length == 1 && resourceId[0] == Separator)
+            {
+                //this is the same as Root but can't return that since this is a ctor
+                Init(null, ResourceType.Tenant, string.Empty, false, null);
+            }
+            else
+            {
+                ReadOnlySpan<char> remaining = resourceId.AsSpan();
 
-        private ResourceIdentifier(ResourceIdentifier parent, string resourceTypeName, string resourceName)
-            : this(parent, ChooseResourceType(resourceTypeName, parent), resourceName, false)
-        {
+                if (!remaining.StartsWith(SubscriptionStart.AsSpan()) && !remaining.StartsWith(ProviderStart.AsSpan()))
+                    throw new ArgumentOutOfRangeException(nameof(resourceId), $"The ResourceIdentifier must start with {SubscriptionStart} or {ProviderStart}.");
+
+                remaining = remaining.Slice(1);
+                //trim trailing '/' off the end if it exists
+                if (remaining[remaining.Length - 1] == Separator)
+                    remaining = remaining.Slice(0, remaining.Length - 1);
```

Comment:
This could be tuned into a single call to `Slice`.

Reply:
That shaved off 5 nanoseconds
|           Method |      Mean |    Error |   StdDev |  Gen 0 |  Gen 1 | Allocated |
|----------------- |----------:|---------:|---------:|-------:|-------:|----------:|
| CreateResourceId | 340.25 ns | 1.810 ns | 1.693 ns | 0.1144 | 0.0005 |     720 B |
|     CreateString |  19.56 ns | 0.228 ns | 0.202 ns | 0.0472 | 0.0000 |     296 B |